policy_agent/policy_agent.py

```diff
@@ -0,0 +1,655 @@
+                           (kind, target_key))
+        elif action == CMD_MODIFIED:
+            if target_key in resource_pool:
+                _log.info("Updating %s %s" % (kind, target_key))
+                resource_pool[target_key] = obj
+            else:
+                _log.warning("Tried to Modify %s %s, but it was not in bin. "
+                             "Treating as Addition" % (kind, target_key))
+                self.process_resource(action=CMD_ADDED,
+                                      kind=kind,
+                                      target=target)
+        else:
+            _log.error("Event in process_resource not recognized: %s" % r_type)
+
+    def delete_resource(kind, target):
+        if kind == KIND_NAMESPACE:
+            resource_pool = self.namespaces
+            obj = Namespace(target)
+
+        elif kind == KIND_SERVICE:
+            resource_pool = self.services
+            obj = Service(target)
+
+        elif kind == KIND_POD:
+            resource_pool = self.pods
+            obj = Pod(target)
+
+        elif kind == KIND_ENDPOINTS:
+            resource_pool = self.endpoints
+            obj = Endpoints(target)
+
+        target_key = obj.key
+
+        if target_key in resource_pool:
+            del resource_pool[target_key]
+            _log.info("%s %s deleted from Calico store" % (kind, target_key))
+
+        else:
+            _log.error("Tried to Delete %s %s but it was not in bin" % \
+                       (kind, target_key))
+
+    def match_pod(self, namespace, name):
+        try:
+            return self.pods["%s/%s" % (namespace, name)]
+        except KeyError:
+            _log.error("Pod %s in NS %s not processed" % (name, namespace))
+            return None
+
+    def resync(self):
+        """
+        Updates Calico store with new resource information
+        """
+        for namespace_key, ns in self.changed_namespaces.items():
+            # Create/update Namespace Profiles
+            if ns.create_ns_profile():
+                self.nsResourceVersion = ns.resourceVersion
+                self.namespaces[namespace_key] = ns
+                del self.changed_namespaces[namespace_key]
+
+        for service_key, svc in self.changed_services.items():
+            # Process Services
+            self.svcResourceVersion = svc.resourceVersion
+            self.services[service_key] = svc
+            del self.changed_services[service_key]
+
+        for pod_key, po in self.changed_pods.items():
+            # Apply Namespace policy to new/updated pods
+            if po.apply_ns_policy():
+                if po.remove_default_profile():
+                    self.poResourceVersion = po.resourceVersion
+                    self.pods[pod_key] = po
+                    del self.changed_pods[pod_key]
+
+        # As endpoint lists change, create/add new profiles to pods as
+        # necessary
+        # TODO: if some object not ready for resync, finish loop w/out moving to
```

Comment:
If the endpoints watcher connects to the API before the pod watcher, the resources will be placed in the queue out of the assumed order. Logically changes to pods and services will trigger a subsequent Endpoints change, but our Threads are pulling updates asynchronously, so who knows what order things show up in queues?

Also resync may not always process everything in one go. If something in pod processing fails, it will skip and retry the next cycle.